tests/window_management_tests/test_minimal_window_manager.cpp

```diff
@@ -0,0 +1,487 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <miral/minimal_window_manager.h>
+#include <mir/scene/session.h>
+#include <mir/wayland/weak.h>
+#include <mir/scene/surface.h>
+#include <mir/executor.h>
+#include <mir/events/event_builders.h>
+#include <mir/events/pointer_event.h>
+#include <mir/compositor/buffer_stream.h>
+#include <mir/graphics/buffer_basic.h>
+#include <mir/test/doubles/stub_buffer_stream.h>
+
+#include "mir_test_framework/window_management_test_harness.h"
+#include <linux/input.h>
+
+namespace ms = mir::scene;
+namespace msh = mir::shell;
+namespace geom = mir::geometry;
+namespace mg = mir::graphics;
+namespace mev = mir::events;
+using namespace testing;
+
+MATCHER_P(LockedEq, value, "")
+{
+    return arg.lock() == value;
+}
+
+class MinimalWindowManagerTest : public mir_test_framework::WindowManagementTestHarness
+{
+public:
+    auto get_builder() -> mir_test_framework::WindowManagementPolicyBuilder override
+    {
+        return [&](miral::WindowManagerTools const& tools)
+        {
+            return std::make_unique<miral::MinimalWindowManager>(tools);
+        };
+    }
+
+    auto get_output_rectangles() -> std::vector<mir::geometry::Rectangle> override
+    {
+        return {
+            mir::geometry::Rectangle{{0, 0}, {800, 600}},
+            mir::geometry::Rectangle{{800, 0}, {800, 600}}
+        };
+    }
+};
+
+TEST_F(MinimalWindowManagerTest, new_window_has_focus)
+{
+    auto const app = open_application("test");
+    msh::SurfaceSpecification spec;
+    spec.width = geom::Width {100};
+    spec.height = geom::Height{100};
+    spec.depth_layer = mir_depth_layer_application;
+    auto window = create_window(app, spec);
+    EXPECT_EQ(focused_surface(), window.operator std::shared_ptr<ms::Surface>());
+}
+
+TEST_F(MinimalWindowManagerTest, alt_f4_closes_active_window)
+{
+    auto const app = open_application("test");
+    msh::SurfaceSpecification spec;
+    spec.width = geom::Width {100};
+    spec.height = geom::Height{100};
+    spec.depth_layer = mir_depth_layer_application;
+    auto window = create_window(app, spec);
+
+    // Process an alt + f4 input
+    std::chrono::nanoseconds const event_timestamp = std::chrono::system_clock::now().time_since_epoch();
+    MirKeyboardAction const action{mir_keyboard_action_down};
+    xkb_keysym_t const keysym{0};
+    int const scan_code{KEY_F4};
+    MirInputEventModifiers const modifiers{mir_input_event_modifier_alt};
+    auto const event = mir::events::make_key_event(
+        mir_input_event_type_key,
+        event_timestamp,
+        action,
+        keysym,
+        scan_code,
+        modifiers);
+    publish_event(*event);
+}
```

Comment:
Where is the check that the window has been closed?

Reply:
True :wink: